---
title: "The Hunter Suite and DEgenesHunter "
author:
-name: "James R. Perkins"
-affiliation: Fred Hutchinson Cancer Research Center, 1100 Fairview Ave. N., P.O. Box 19024, Seattle, WA, USA 98109–1024
date: "`r format(Sys.Date(), '%m/%d/%Y')`"
output:
  rmarkdown::html_document:
    highlight: pygments
    toc: true
    fig_width: 5
vignette: >
  %\VignetteIndexEntry{Analyzing RNA-seq data with DESeq2}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
  %\usepackage[utf8]{inputenc}
---



# Version Info
```{r, echo=FALSE, results="hide", warning=FALSE}
suppressPackageStartupMessages({library('DEgenesHunter')})
```

 <p>
 **R version**: `r R.version.string`
 <br />
 **Bioconductor version**: `r BiocManager::version()`
 <br />
 **Package**: `r packageVersion("DEgenesHunter")`
 <\p>


# Introduction

DEgenes Hunter implements a comprehensive protocol for the analysis of transcriptional data. It covers all key steps in DEG detection, CEG detection and functional analysis for RNA-seq data. It is presented as an R package containing functions that can be run interactively, it also contains command line scripts that allow it to be run on computer clusters, for example.

# Standard Package Usage

## Differential Expression Analysis

The most basic use of the package is to perform DE gene analysis through the use of previously published methods. DEgenes Hunter will, following some initial preprocessing, run the different methods, combine the results and produce an output report, as well as an single output table containing the results of all of the methods used, and their combined scores.

The combined scores consist of the combined FC values, calculated by XXX, and the combined adjusted p-value (FDR) values, calculated by XXX.

To use DEgenes hunter with only a single package, effectively as a wrapper, one can run the following command:


```{r standard_DEA_one_pack, eval=FALSE}
data(toc)
data(target)
degh_out_one_pack <- main_degenes_Hunter(raw=toc, target=target, modules="D")
```

Where toc and target are ...

To use it with multiple packages, one can run the following command:

```{r standard_DEA_multi_pack, eval=TRUE}
degh_out_multi_pack <- main_degenes_Hunter(raw=toc, target=target, modules="DEL")
```

```{r standard_DEA_results, eval=TRUE}
head(degh_out_multi_pack$DE_all_genes)
```

The output is a list, which includes, in slot *DE_all_genes* a data.frame containing details of the different DE packages implemented, : 

PREVALENT_DEGS are those that are considered significant in at least n algorithms.
POSSIBLE_DEGS are those considered significant in at least 1 method. Clearly, PREVALENT_DEGS and POSSIBLE_DEGS will be the same in the case that n = 1.

To be considered significant for a given method, a gene must have an adjusted-pvalue < XX and a logFC > XX , these values are adjustable using the arguments XX and XX. For example, for a more relaxed experiment, one could run the function:

XXXXX

It also includes NOT_DEGS and FILTERED_OUT

There is another column, combined XXX – which is POS/NEG depending on whether the combined p-value calculated by combining the methods is <= the value of XXX (0.05, as described above).

The report generated by this step will be generated within the output folder with name. 

XXX Details of reports XXX

More complex model designs.

In order to control for specific variables (such as individuals in paired designs, potential confounding factors such as age, etc.), 

For example, in an experimental design where the same samples receive treatment, we can use a paired design:

SHOW TOP OF THE TOC

SHOW TARGET

SHOW HOW TO ANALYZE THE DATA


```{r standard_DEA_model, eval=FALSE}
```


`



