---
author: "SysBioLab"
output:
  html_document:
    toc: true
    toc_float: true
    df_print: paged
    fig_width: 12
---

<style type='text/css'>
    body .main-container { 
        max-width: 90%;
    
    }

    .plot_real_size {
      overflow: scroll;
      max-height: 600px;
    }

    .plot_real_size img{
      max-width: none;
      max-height: none;
    }

    embed {
   
  }   
</style>


# **CoRmiT RNAseq - miRNAseq correlation report**

This report summarizes ExpHunterSuite CoRmiT results comparing RNAseq and miRNAseq data and looking for putative miRNA-gene target pairs, 
representing genes that are silenced by miRNAs.

## **Strategies comparison**

miRNAseq and RNAseq data are compared and correlated using multiple strategies and then compared with multiMiR databases.

* Description of strategies used: 
```{r print_info, echo = FALSE, warning = FALSE, results = "asis"}
  sig_strategies <- names(strategies)[! names(strategies) %in% unsig_strategies]
  cat(parse_strat_text(sig_strategies))

```

* Three control strategies are also used:
    + **DEGs_DEMs_permutated:** All possible permutations between differentially expressed genes and differentially expressed miRNAs. Correlation are not performed.
    + **Eigengene_0_RNA_vs_miRNA_normalized_counts:** Correlation between the Eigengene profile for gene module 0, which contains genes that could not be assigned to any other consistent module and miRNA normalizaed counts.
    + **normalized_counts_RNA_vs_miRNA_Eigengene_0:** Correlation between gene normalized counts and the Eigengene profile for miRNA module 0, which contains miRNAs that could not be assigned to any other consistent module.


Please note that graphics are only displayed for the strategies that were run and for which significant results were found.

```{r sig_unsig, echo = FALSE, warning = FALSE, results = "asis"}

  sig_strats <-  paste0("**", paste(sig_strategies, collapse = "**, **"), "**")

  unsig_strats <- paste0("**", paste(unsig_strategies, collapse = "**, **"), "**")

 cat(paste0("In this report, only the strategies ", sig_strats, " produced signficant results.\n"))
 if (length(unsig_strategies) > 0){
  cat(paste0(unsig_strats, " were run, however no signficant correlated pairs were obtained."))
 } 

```

### **Correlation distributions**

Distributions of correlation values between RNAseq and miRNAseq data obtained using the different strategies.
Dashed line shows the correlation cut-off below which miRNA-target pairs are considered significant. 
Horizontal lines within the distributions represent the 90 and 95 percentiles. 
Note that **DEGs_DEMs_permutated** strategy is not included.


```{r methods_comparison_violin_cor, echo=FALSE, results = 'asis', warning = FALSE, message = FALSE}

pp <- ggplot2::ggplot(all_cor_dist, ggplot2::aes(x = strategy, y = correlation, fill = strategy)) +
  ggplot2::geom_violin(col = "white",trim = FALSE, scale = "width",draw_quantiles = c( 0.1, 0.05)) + #ggplot2::ylim(-1.000001, min + 0.00001) +
  ggplot2::geom_hline(yintercept= corr_cutoff, linetype = "dashed", colour = "black") +
  ggplot2::geom_boxplot(width=0.05)+ ggplot2::ylim(-1, 1)+
  ggplot2::coord_flip()+
  ggplot2::theme(axis.text.y = ggplot2::element_blank(), axis.title.y = ggplot2::element_blank()) +
  ggplot2::labs(fill = "Strategy")

pp 
```

### **Pearson's correlation P-values distribution**

Distributions of correlation P-values between RNAseq and miRNAseq data obtained using the different strategies.
Note that **DEGs_DEMs_permutated** strategy is not included.

```{r methods_comparison_violin_p, echo=FALSE, results = 'asis', warning = FALSE, message = FALSE}

pp <- ggplot2::ggplot(all_cor_dist, ggplot2::aes(x = strategy, y = pval, fill = strategy)) +
  ggplot2::geom_violin(trim = FALSE, scale = "width") + 
  ggplot2::geom_hline(yintercept= p_val_cutoff, linetype = "dashed", colour = "black") +
  ggplot2::geom_boxplot(width=0.05)+ ggplot2::ylim(-0.01, 1) +
  ggplot2::coord_flip()+
  ggplot2::theme(axis.text.y = ggplot2::element_blank(), axis.title.y = ggplot2::element_blank()) +
  ggplot2::labs(fill = "Strategy")
pp

```

### **Strategies overlapping**

This table shows the number of significant pairs overlaped between strategies.



```{r methods_comparison_intersection, echo=FALSE, results = 'asis', warning = FALSE, message = FALSE}
 subchunk_count <- 3
  common_pairs <- sig_pairs[["all"]]
  sig_pairs[["all"]] <- NULL
  common_pairs <- as.data.frame(as.table(as.matrix(common_pairs)))
  names(common_pairs) <- c("x_axis", "y_axis",  "intersection")
  common_pairs$text <- as.character(common_pairs$intersection)
  common_pairs$x_axis <- as.character(common_pairs$x_axis)
  common_pairs$y_axis <- as.character(common_pairs$y_axis)
  common_pairs[common_pairs$x_axis == common_pairs$y_axis, "text"] <- "-"
  pp <- gg_heatmap(data =  common_pairs, 
              x_axis = "x_axis",
              y_axis= "y_axis",
              fill = "intersection",
              text_plot= "text", 
              labs = FALSE)
  subchunk_count <- rechunk(pp, counter = subchunk_count, 
          chunk_options = "echo=FALSE, results = 'asis', warning = FALSE, message = FALSE")

```

### **Strategies overlapping by group databases**

Following tables shows the number of significant pairs included on each group of multiMiR database overlaped between strategies.

MultiMiR databases can be classified based on those containing predicted pairs and experimentally validated pairs.
We therefore repeated the steps indicated above for the union of **predicted** databases, **validated** databases, and intersection of them **(pred_and_val)**.


```{r methods_comparison_intersection_groups, echo=FALSE, results = 'asis', warning = FALSE, message = FALSE}

for (db_group in names(sig_pairs)){
   common_pairs <- sig_pairs[[db_group]]
   check_ovlp <- common_pairs
   diag(check_ovlp) <- 0
  cat(paste0("#### **Overlap of significant pairs found on ", db_group, " databases**\n\n"))
   if (any(check_ovlp > 0)){
       common_pairs <- as.data.frame(as.table(as.matrix(common_pairs)))
       names(common_pairs) <- c("x_axis", "y_axis",  "intersection")
       common_pairs$text <- as.character(common_pairs$intersection)
       common_pairs$x_axis <- as.character(common_pairs$x_axis)
       common_pairs$y_axis <- as.character(common_pairs$y_axis)
       common_pairs[common_pairs$x_axis == common_pairs$y_axis, "text"] <- "-"
       pp <- gg_heatmap(data =  common_pairs, 
                   x_axis = "x_axis",
                   y_axis= "y_axis",
                   fill = "intersection",
                   text_plot= "text", 
                   labs = FALSE)
       subchunk_count <- rechunk(pp, counter = subchunk_count, 
               chunk_options = "echo=FALSE, results = 'asis', warning = FALSE, message = FALSE")
   } else {
       cat(paste0("There were not overlap between significant pairs of strategies on ", db_group, " databases.\n\n"))
   }
}
   
```

### **Overlap between strategy results and databases**

* This bar chart shows, for each strategy, how many significant miRNA-target pairs:
  + Include a putative **novel** miRNA
  + Include a **known** miRNA (present in miRBASE)

It also shows the number of significant pairs involving known miRNAs that are found in gorups of **multiMiR** databases.
To assess the significance of this number, we select **random** pairs of expressed genes/miRNAs from the 
experiment, of an equivalent number to the significant pairs including a known miRNA, and note how many of these
pairs overlap with the multiMiR databases. This is repeated multiple times, and the mean +/- standard deviation are plotted.



```{r methods_comparison_bar, echo=FALSE, results = 'asis', warning = FALSE, message = FALSE}

pp <- ggplot2::ggplot(filters_summary, ggplot2::aes(y = pairs, x = strategy, fill = type)) + 
      ggplot2::geom_bar(stat= "identity",position=ggplot2::position_dodge()) +
      ggplot2::geom_errorbar(ggplot2::aes(ymin=pairs-sdev, ymax=pairs+sdev), width=.2,
                 position=ggplot2::position_dodge(.9)) + 
      ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 20, hjust = 1)) + 
      ggplot2::geom_text(ggplot2::aes(label = pairs), angle = 45, position = ggplot2::position_dodge(width=0.95), size=2, vjust = -2)
pp
```


### **Association with databases**

* **Statistical metrics:**
  + **Accuracy**, **Precision** and **Specificity** were calculated for the overlapping between significant strategies and group of multiMiR databases.
  + **Fisher exact test significance (-LogPval):** Significant and non-significant pairs were compared in terms of overlap with databases using one-tailed Fisher exact test. P-values are shown for the different strategies and for the different databases considered. -log10(P-values) are represented. Radar plot shows -log10(0.05) threshold with a dashed grey line.
  + **Likelihood ratio for positive test (LRplus_test):** For each pair, this shows to what extent a significant result indicates an increased probability of being a true miRNA-target pair. It is calculated as the True Positive Rate, calculated as proportion of all pairs in the experiment that overlap with the database that are signficant, divided by the False Positive Rate, calculated as proportion of all pairs in the experiment that do not overlap with the database that are significant.
  + **Likelihood ratio for positive subjects (LRplus_sub):** Given a true miRNA-target pair, what is the increased probability of obtaining a significant result. It is calculated as the proportion of significant results overlapping with the databases (True positive rate), divided by the proportion of non-signficant results overlapping with the databases (false positive rate).

*(Radar plots show a threshold of 1 for the Likelihood ratios)*


```{r prec_rec, echo=FALSE, results = 'asis', warning = FALSE, message = FALSE}

    cont_table_radar <- cont_tables[c("strategy","Accuracy","Precision","Specificity","db_group", "Pvalue", "LRplus_test", "LRplus_sub")]
    cont_table_radar[,"-LogPval"] <- -log10(cont_table_radar$Pvalue)
    cont_table_radar$Pvalue <- NULL
    cont_table_radar <- split(cont_table_radar, cont_table_radar$db_group)
    thresholds <- data.frame(check.names = FALSE, row.names = "Thresholds", Accuracy = 0, Precision = 0, Specificity = 0, "-LogPval" = -log10(0.05), LRplus_test = 1, LRplus_sub = 1)
    for(db_group in c("multimir", "predicted","validated", "pred_and_val")){
      cont_tab <- cont_table_radar[[db_group]]
      cat(paste0("\n\n#### **", stringr::str_to_title(db_group), " databases**\n\n"))
      cat(paste0("\n\nThis section sumarizes all metrics computed for each strategy compared with ", db_group, " databases.\n\n"))
      rownames(cont_tab) <- cont_tab$strategy
      cont_tab$db_group <- NULL
      cont_tab$strategy <- NULL
      COL<- grDevices::colorRampPalette(c("red", "blue", "green", "orange"))(nrow(cont_tab))
      COL <- c(COL, "#9C9C9C")
      scales <- matrix(NA,nrow = 2, ncol = ncol(cont_tab) )
      cont_tab_raw <- cont_tab
      linetypes <-c(rep(1,nrow(cont_tab)), 2)
      cont_tab <- rbind(cont_tab, thresholds)
      scales[1,] <- c(1,1,1, max(cont_tab$LRplus_test), max(cont_tab$LRplus_sub), max(cont_tab[,"-LogPval"]))
      scales[2,] <- c(0,0,0, min(cont_tab$LRplus_test), min(cont_tab$LRplus_sub), min(cont_tab[,"-LogPval"]))
      scales <- as.data.frame(scales)
      dimnames(scales) <- list(c("Max","Min"), colnames(cont_tab))
      cont_tab_lim <- rbind(scales, cont_tab)
          par(mfrow = c(1, 2))      
          fmsb::radarchart(cont_tab_lim, cglcol="grey", maxmin = TRUE, pcol = COL, plty=linetypes )
          plot.new() 
          graphics::legend("center", legend = rownames(cont_tab), title = "Strategy", 
            col = COL, seg.len = 2, border = "transparent", pch = 16, lty = linetypes)
      if(db_group =="multimir"){
            p <- recordPlot()
            subchunk_count <- rechunk(p,
                         counter = subchunk_count, 
                         chunk_options = "echo=FALSE, results = 'asis', warning = FALSE, message = FALSE")
      } 
      cont_tab$db_group <- NULL
      subchunk_count <- rechunk(DT::datatable(cont_tab_raw, filter = 'top', rownames = TRUE, extensions = c('Buttons','ColReorder'),
                                    options = list( paging =TRUE, colReorder = TRUE, dom = 'lftBip',  buttons = c('copy', 'csv', 'excel'))),
                          counter = subchunk_count, 
                          chunk_options = "echo=FALSE, results = 'asis', warning = FALSE, message = FALSE")
    }     

```

### **Prediction scores for the significant pairs**


The pairs in the multiMiR prediction databases are assigned scores that represent confidence in the predicted miRNA-target interaction.  

For each prediction database, the scores for the significant pairs obtained by each strategy were compared with random subsets of pairs.

The comparison was performed using one-tailed T-tests and one-tailed bootstrap T-tests. P-value for both methods are included.

Lower P-values means that the significant pairs had higher scores than the random subests.

The final column of the table shows the combined p-values for all databases, integrated using Fishers method.

#### **One-tailed T-test**

```{r score_comp_pval, results = 'asis', echo = FALSE, warning = FALSE, message = FALSE}
score_comp_pval <- reshape(score_comp[,c("strategy","database","p.value")], idvar = "strategy", timevar = "database", direction = "wide")
    colnames(score_comp_pval) <- gsub("p.value.", "",colnames(score_comp_pval))
   DT::datatable(score_comp_pval, filter = 'top', rownames = FALSE, extensions = c('Buttons','ColReorder'),
            options = list(
              paging = FALSE,
              colReorder = TRUE,
              dom = 'lftBip',
              buttons = c('copy', 'csv', 'excel')
          ))

```

#### **One-tailed bootstrap T-test**

```{r score_comp_bootpval, results = 'asis', echo = FALSE, warning = FALSE, message = FALSE}
  score_comp_boot_pval <- reshape(score_comp[,c("strategy","database","boot.p.value")], idvar = "strategy", timevar = "database", direction = "wide")
   colnames(score_comp_boot_pval) <- gsub("boot.p.value.", "",colnames(score_comp_boot_pval))
  DT::datatable(score_comp_boot_pval, filter = 'top', rownames = TRUE, extensions = c('Buttons','ColReorder'),
            options = list(
              paging = FALSE,
              colReorder = TRUE,
              dom = 'lftBip',
              buttons = c('copy', 'csv', 'excel')
          ))

```

#### **Comparison of p-values obtained between T-test and boostrap T-test**

This plot shows -log(P) values from the T-test and bootstrap T-test. Grey dotted line indicates -log(0.05).

```{r score_comp_pval_log, results = 'asis', echo = FALSE, warning = FALSE, message = FALSE}
    log_score_comp_boot <- score_comp[,c("strategy","database","log.boot.p.value")]
    data.table::setnames(log_score_comp_boot, c("log.boot.p.value"), c("log.p.value"))
    log_score_comp_boot$test_type <- "boot.t.test"
    log_score_comp <- score_comp[,c("strategy","database","log.p.value")]
    log_score_comp$test_type <- "t.test"
    log_score_comp <- rbind(log_score_comp_boot, log_score_comp)

  ggplot2::ggplot(log_score_comp, ggplot2::aes(x = database, y = log.p.value, color = strategy, group = paste0(strategy, test_type))) +
  ggplot2::geom_line(ggplot2::aes(linetype = test_type)) + 
  ggplot2::geom_line(ggplot2::aes( y =  -log(0.05), linetype = "-log(0.05)" ), colour = "grey") +
  # ggplot2::geom_point(size = 1) +
  ggplot2::labs(linetype = "T test type", colour = "Strategy", group = "Strategy", shape = "T test type") +
  ggplot2::scale_linetype_manual(values=c("dotted", "dashed", "solid"))+
  ggplot2::theme(legend.position = "bottom", legend.box = "vertical")
```


## **Strategies specific results**

```{r render_partial_reports, echo = FALSE, warning = FALSE}
  curr_out <- ""

  curr_out <- unlist(lapply(sig_strategies[sig_strategies != "DEGs_DEMs_permutated"] ,function(method_name){
      sec_text <- paste(c("### **", method_name ," correlation method**\n"),collapse="") # Header
      sec_text <- c(sec_text, knitr::knit_expand("partial/miRNA_cor_methods.Rmd")) # Body
      return(sec_text)
  }))    

``` 
`r paste(knitr::knit(text = curr_out), collapse = '\n')`


## **Score distribution sampling check**
These plots compare the whole score distribution of background pairs with his subset. If the distributions are very different you have to increase the sample_proportion option.


```{r score_sampling, results = "asis", echo = FALSE, warning = FALSE}
  for (database in selected_predicted_databases) {
    background_score <- all_pairs[,get(database)]

    # save(background_score, file = "/home/bio267lab/proyectos/target_miRNA_2020/test2.Rdata")
    background_score <- background_score[!is.na(background_score)]
    background_score_sample <- background_score[sample(length(background_score), round(length(background_score) * sample_proportion), replace = FALSE)]
    
    background_score <- data.frame(background_score)
    colnames(background_score) <- database
    background_score_sample <- data.frame(background_score_sample)
    colnames(background_score_sample) <- database
    gridExtra::grid.arrange (
    
      ggplot2::ggplot(background_score,        ggplot2::aes_(x = as.name(database))) + 
                                               ggplot2::geom_density() + 
                                               ggplot2::ggtitle("Complete score distribution")
                                          ,
      ggplot2::ggplot(background_score_sample, ggplot2::aes_(x = as.name(database))) + 
                                               ggplot2::geom_density() + 
                                               ggplot2::ggtitle("Sampled score distribution"),
      ncol = 2,
      top = grid::textGrob(stringr::str_to_title(database),gp=grid::gpar(fontsize=20,font=3)))
       
    }

  cat("\n\n")
```




