---
author: "SysBioLab"
output:
  html_document:
    toc: true
    toc_float: true
    df_print: paged
    fig_width: 12
---

<style type='text/css'>
    body .main-container { 
        max-width: 90%;
    
    }

    .plot_real_size {
      overflow: scroll;
      max-height: 600px;
    }

    .plot_real_size img{
      max-width: none;
      max-height: none;
    }

    embed {
   
  }   
</style>


# **CoRmiT: RNAseq - miRNAseq correlation report**

This report summarizes the results of ExpHunterSuite CoRmiT, comparing RNAseq and miRNAseq data and looking for putative miRNA-gene target interactions based on different strategies.

## **Strategies comparison**

miRNAseq and RNAseq data are compared and expression levels correlated using multiple strategies. 
Correlated miRNA-mRNA pairs are known **significant pairs** representing putative miRNA-mRNA target interactions. 
Results are then compared with multiMiR databases.

* Brief description of strategies used: 
```{r print_info, echo = FALSE, warning = FALSE, results = "asis"}
  sig_strategies <- names(strategies)[! names(strategies) %in% unsig_strategies]
  cat(parse_strat_text(sig_strategies))

  mock_strategies <-   c("Eigengene_0_RNA_vs_miRNA_normalized_counts", 
      # "normalized_counts_RNA_vs_miRNA_Eigengene_0", 
      # "DEGs_RNA_vs_miRNA_DEMs",
      "DEGs_DEMs_permutated")

```

* Three **control strategies** are also used:
    + **DEGs_DEMs_permutated:** All possible permutations between differentially expressed genes and differentially expressed miRNAs. Correlation is not performed.
    + **Eigengene_0_RNA_vs_miRNA_normalized_counts:** Correlation between the Eigengene profile for gene module 0, which contains genes that could not be assigned to any other module and miRNA normalized counts.

Graphics are only displayed for the strategies that were run and for which significant results were found. 
The **control strategies** are artifical strategies used for comparison with the real strategies in terms of performance.
Their results should not be considered for miRNA-gene target selection.

```{r sig_unsig, echo = FALSE, warning = FALSE, results = "asis"}

    # + **normalized_counts_RNA_vs_miRNA_Eigengene_0:** Correlation between gene normalized counts and the Eigengene profile for miRNA module 0, which contains miRNAs that could not be assigned to any other module.
  sig_strats <-  paste0("**", paste(sig_strategies, collapse = "**, **"), "**")

  unsig_strats <- paste0("**", paste(unsig_strategies, collapse = "**, **"), "**")

 cat(paste0("In this report, only the strategies ", sig_strats[!sig_strats %in% mock_strategies], " produced signficant results.\n\n"))
 if (length(unsig_strategies) > 0){
  cat(paste0(unsig_strats, " were run, however no signficant correlated pairs were obtained."))
 } 

```

### **Correlation distributions**

Distributions of correlation values between miRNAs and target genes obtained using the different strategies.
Dashed line shows the correlation cut-off below which miRNA-target pairs are considered significant. 
Horizontal white lines within the distributions represent the 90 and 95 percentiles. 
The **DEGs_DEMs_permutated** strategy is not included.


```{r methods_comparison_violin_cor, echo=FALSE, results = 'asis', warning = FALSE, message = FALSE}

pp <- ggplot2::ggplot(all_cor_dist, ggplot2::aes(x = strategy, y = correlation, fill = strategy)) +
  ggplot2::geom_violin(col = "white",trim = FALSE, scale = "width",draw_quantiles = c( 0.1, 0.05)) + #ggplot2::ylim(-1.000001, min + 0.00001) +
  ggplot2::geom_hline(yintercept= corr_cutoff, linetype = "dashed", colour = "black") +
  ggplot2::geom_boxplot(width=0.05)+ ggplot2::ylim(-1, 1)+
  ggplot2::coord_flip()+
  ggplot2::theme(axis.text.y = ggplot2::element_blank(), axis.title.y = ggplot2::element_blank()) +
  ggplot2::labs(fill = "Strategy")

pp 
```

### **Pearson's correlation p-values distribution**

Distributions of correlation p-values between miRNAs and target genes obtained using the different strategies.
The dashed line represents a p-value of 0.05
The **DEGs_DEMs_permutated** strategy is not included.

```{r methods_comparison_violin_p, echo=FALSE, results = 'asis', warning = FALSE, message = FALSE}

pp <- ggplot2::ggplot(all_cor_dist, ggplot2::aes(x = strategy, y = pval, fill = strategy)) +
  ggplot2::geom_violin(trim = FALSE, scale = "width") + 
  ggplot2::geom_hline(yintercept= p_val_cutoff, linetype = "dashed", colour = "black") +
  ggplot2::geom_boxplot(width=0.05)+ ggplot2::ylim(-0.01, 1) +
  ggplot2::coord_flip()+
  ggplot2::theme(axis.text.y = ggplot2::element_blank(), axis.title.y = ggplot2::element_blank()) +
  ggplot2::labs(fill = "Strategy")
pp

```

### **Overlapping strategies**

This table shows the number of **significant pairs** overlapping between strategies.


```{r methods_comparison_intersection, echo=FALSE, results = 'asis', warning = FALSE, message = FALSE}
 subchunk_count <- 3
  common_pairs <- sig_pairs[["all"]]
  sig_pairs[["all"]] <- NULL
  common_pairs <- as.data.frame(as.table(as.matrix(common_pairs)))
  names(common_pairs) <- c("x_axis", "y_axis",  "intersection")
  common_pairs$text <- as.character(common_pairs$intersection)
  common_pairs$x_axis <- as.character(common_pairs$x_axis)
  common_pairs$y_axis <- as.character(common_pairs$y_axis)
  # common_pairs[common_pairs$x_axis == common_pairs$y_axis, "text"] <- "-"
  pp <- gg_heatmap(data =  common_pairs, 
              x_axis = "x_axis",
              y_axis= "y_axis",
              fill = "intersection",
              text_plot= "text", 
              labs = FALSE)
  subchunk_count <- rechunk(pp, counter = subchunk_count, 
          chunk_options = "echo=FALSE, results = 'asis', warning = FALSE, message = FALSE")

```

### **Overlapping strategies for different database groups**

The following tables shows the number of significant pairs that were found in multiMiR database groups for each strategy, and their overlap.

MultiMiR pairs can be classified based on which kind of database they appear in. The two main groups are **predicted** pairs and experimentally **validated** pairs.
We therefore repeated the steps indicated above for the union of **predicted** databases, **validated** databases, and their intersection **(pred_and_val)**.


```{r methods_comparison_intersection_groups, echo=FALSE, results = 'asis', warning = FALSE, message = FALSE}

for (db_group in names(sig_pairs)){
   common_pairs <- sig_pairs[[db_group]]
   check_ovlp <- common_pairs
   diag(check_ovlp) <- 0
  cat(paste0("#### **Overlap of significant pairs found on ", stringr::str_to_title(db_group), " databases**\n\n"))
   if (any(check_ovlp > 0)){
       common_pairs <- as.data.frame(as.table(as.matrix(common_pairs)))
       names(common_pairs) <- c("x_axis", "y_axis",  "intersection")
       common_pairs$text <- as.character(common_pairs$intersection)
       common_pairs$x_axis <- as.character(common_pairs$x_axis)
       common_pairs$y_axis <- as.character(common_pairs$y_axis)
       # common_pairs[common_pairs$x_axis == common_pairs$y_axis, "text"] <- "-"
       pp <- gg_heatmap(data =  common_pairs, 
                   x_axis = "x_axis",
                   y_axis= "y_axis",
                   fill = "intersection",
                   text_plot= "text", 
                   labs = FALSE)
       subchunk_count <- rechunk(pp, counter = subchunk_count, 
               chunk_options = "echo=FALSE, results = 'asis', warning = FALSE, message = FALSE")
   } else {
       cat(paste0("There were not overlap between significant pairs of strategies for ", db_group, " databases.\n\n"))
   }
}
   
```

### **Comparison of the overlap between strategy results and databases**

* This bar chart shows, for each strategy, how many significant miRNA-target pairs:
  + Include a putative **novel** miRNA
  + Include a **known** miRNA (present in miRBASE)

It also shows the number of significant pairs involving known miRNAs that are found in the different groups of multiMiR databases.
To put these numbers in context, we also select **random** pairs of expressed genes/miRNAs from the 
experiment, of an equivalent number to the significant pairs including a known miRNA, and note how many of these
pairs overlap with the multiMiR databases. This is repeated multiple times, and the mean +/- standard deviation are plotted.



```{r methods_comparison_bar, echo=FALSE, results = 'asis', warning = FALSE, message = FALSE}

pp <- ggplot2::ggplot(filters_summary, ggplot2::aes(y = pairs, x = strategy, fill = type)) + 
      ggplot2::geom_bar(stat= "identity",position=ggplot2::position_dodge()) +
      ggplot2::geom_errorbar(ggplot2::aes(ymin=pairs-sdev, ymax=pairs+sdev), width=.2,
                 position=ggplot2::position_dodge(.9)) + 
      ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 20, hjust = 1)) + 
      ggplot2::geom_text(ggplot2::aes(label = pairs), angle = 45, position = ggplot2::position_dodge(width=0.95), size=2, vjust = -2)
pp
```

### **Significance of the overlap**

Significant and non-significant pairs were compared in terms of overlap with databases using one-tailed Fisher exact test. 
For prediction databases the test were performed idependently for each prediction database. The p-values were then combined using Fisher combined probability test (**predicted**). 
Pairs found in the validation databases (**validated**) and validated pairs that are also predicted by any database (**pred_and_val**) were grouped together. 
P-values are shown for the different strategies and for the different group of databases. 
-log10(P-values) are shown. The -log10(0.05) threshold is represented as a dashed grey line. 


```{r fisher, echo=FALSE, results = 'asis', warning = FALSE, message = FALSE}
    fisher_test_raw <- cont_tables[c("strategy", "Pvalue", "db_group")]
    comb_pred_pval <- p_fisher[p_fisher$database == "comb_pval",]

    fisher_test_raw <- fisher_test_raw[fisher_test_raw$db_group != "predicted", ]
    fisher_test_raw <- rbind(fisher_test_raw, 
      data.frame(strategy = comb_pred_pval$strategy,
                  Pvalue = comb_pred_pval$fisher.p.value,
                  db_group = "predicted"))
    fisher_test_table <- fisher_test_raw
    fisher_test_table[,"LogPval"] <- -log10(fisher_test_table$Pvalue)

    fisher_test_table$db_group <- factor(fisher_test_table$db_group, c("predicted","validated", "pred_and_val"))
    # fisher_test$group_db <- rownames(fisher_test)
    ggplot2::ggplot(fisher_test_table, ggplot2::aes(x = db_group, y = LogPval, group = strategy,color = strategy)) +
    ggplot2::geom_line() + 
    ggplot2::geom_line(ggplot2::aes( y =  -log10(0.05), linetype = "-log10(0.05)") , colour = "grey") +
    ggplot2::geom_point(size = 1) +
    ggplot2::labs(colour = "Strategy", linetype = "-log10(0.05)", y = "-log10(p)", x = "Pairs group") +
    ggplot2::scale_linetype_manual(values=c("dashed", "solid"))+
    ggplot2::theme(legend.position = "bottom", legend.box = "vertical")
    strat_fisher_sig <- cont_tables %>% dplyr::group_by(strategy) %>% dplyr::summarise(significant = all(Pvalue <= 0.05))
    unsig_strategies <- c(unsig_strategies, strat_fisher_sig[!strat_fisher_sig$significant, "strategy"])
    sig_strategies <- names(strategies)[! names(strategies) %in% unsig_strategies]

```

#### **Comparison of predicted and validated groups**

This plot compares the p-values of the Fisher exact tests performed for each strategy for the **predicted** and **validated** pairs. 
-Log10(p-value) is shown on each axis.
-Log10(0.05) is represented as a black dashed line.


```{r fisher_scatter, echo=FALSE, results = 'asis', warning = FALSE, message = FALSE}

   fisher_test_scatter <- fisher_test_table[,c("strategy","db_group","LogPval")]
   fisher_test_scatter <- tidyr::spread(fisher_test_scatter, db_group, LogPval)
   
   ggplot2::ggplot(fisher_test_scatter,ggplot2::aes(y = predicted, x = validated, fill = strategy, color = strategy)) + 
   ggplot2::geom_point()+
   ggplot2::geom_hline(yintercept = -log10(0.05), linetype = "dashed") + 
   ggplot2::geom_vline(xintercept = -log10(0.05), linetype = "dashed") +
   ggplot2::labs(x = "Validated -log10(p)", y = "Predicted -log10(p)", color = "Strategy", fill = "Strategy")


```

#### **Fisher exact test p-values table**

```{r fisher_table, echo=FALSE, results = 'asis', warning = FALSE, message = FALSE}
  fisher_test_raw <- tidyr::spread(fisher_test_raw, db_group, Pvalue)
  DT::datatable(fisher_test_raw, filter = 'top', rownames = FALSE, extensions = c('Buttons','ColReorder'),
          options = list(
            paging = FALSE,
            colReorder = TRUE,
            dom = 'lftBip',
            buttons = c('copy', 'csv', 'excel')
  ))
```
**NOTE:** Only strategies with significant p-values in all tests will be used for the following plots.


### **Odds ratios for the different strategies**


The plot shows the significant strategies ordered by Odds ratio (OR). 
OR can be thought of as a measure of the association between a given strategy and pairs in the databases. 
OR > 1 represents association while OR = 1 means independence.
The color represents the p-value from the Fisher exact test.
The X-axis represents the OR and the dot size the number of pairs associated with the group of databases.

```{r dotplot, echo=FALSE, results = 'asis', warning = FALSE, message = FALSE}

cont_tables <- cont_tables[! cont_tables$strategy %in% unsig_strategies,]
splitted_c_tables <- split(cont_tables, cont_tables$db_group)
for (db_group in c("predicted", "validated")){
  cat(paste0("#### Comparison with ", stringr::str_to_title(db_group), " pairs\n\n"))
  filtered_cont <- splitted_c_tables[[db_group]]
  idx <- order(filtered_cont$Odds_ratio, decreasing = TRUE)
  filtered_cont$strategy <- factor(filtered_cont$strategy,
                            levels=rev(unique(filtered_cont$strategy[idx])))

  pp <- ggplot2::ggplot(filtered_cont, ggplot2::aes(x = Odds_ratio, y = strategy)) +
  ggplot2::geom_point(ggplot2::aes(color = Pvalue, size = TP)) + 
  ggrepel::geom_text_repel(ggplot2::aes(label = as.character(TP))) +
  ggplot2::scale_color_continuous(low="red", high="blue", name = "Pvalue", guide=ggplot2::guide_colorbar(reverse=TRUE))+
  ggplot2::guides(size = FALSE)
  
  plot(pp)
  cat("\n\n")
}


```


```{r prec_rec, echo=FALSE, results = 'asis', warning = FALSE, message = FALSE, eval = FALSE}
cat("

### **Association with databases**

* **Statistical metrics:**
  + **Accuracy**, **Precision** and **Specificity** were calculated based on the overlap between significant strategies and group of multiMiR databases.
  + **Likelihood ratio for positive test (LRplus_test):** For each pair, this shows to what extent a significant result indicates an increased probability of being a true miRNA-target pair. It is calculated as the True Positive Rate, calculated as proportion of all pairs in the experiment that overlap with the database that are signficant, divided by the False Positive Rate, calculated as proportion of all pairs in the experiment that do not overlap with the database that are significant.
  + **Likelihood ratio for positive subjects (LRplus_sub):** Given a true miRNA-target pair, what is the increased probability of obtaining a significant result. It is calculated as the proportion of significant results overlapping with the databases (True positive rate), divided by the proportion of non-signficant results overlapping with the databases (false positive rate).

*(Threshold dashed grey line marks values of 1 for the Likelihood ratiosand values of 0 for Accuracy, Precission and Specificity)*

  
  ")

    cont_table_radar <- cont_tables[c("strategy","Accuracy","Precision","Specificity","db_group", "LRplus_test", "LRplus_sub")]
    # cont_table_radar[,"-LogPval"] <- -log10(cont_table_radar$Pvalue)
    # cont_table_radar$Pvalue <- NULL
    cont_table_radar <- split(cont_table_radar, cont_table_radar$db_group)
    thresholds <- data.frame(check.names = FALSE, row.names = "Thresholds", Accuracy = 0, Precision = 0, Specificity = 0, LRplus_test = 1, LRplus_sub = 1)
    for(db_group in c("multimir", "predicted","validated", "pred_and_val")){
      cont_tab <- cont_table_radar[[db_group]]
      cat(paste0("\n\n#### **", stringr::str_to_title(db_group), " databases**\n\n"))
      cat(paste0("\n\nThis section sumarizes all metrics computed for each strategy compared with ", db_group, " databases.\n\n"))
      rownames(cont_tab) <- cont_tab$strategy
      cont_tab$db_group <- NULL
      cont_tab_raw <- cont_tab
      cont_tab$strategy <- NULL
      COL<- grDevices::colorRampPalette(c("red", "blue", "green", "orange"))(nrow(cont_tab))
      COL <- c(COL, "#9C9C9C")
      scales <- matrix(NA,nrow = 2, ncol = ncol(cont_tab) )
      linetypes <-c(rep(1,nrow(cont_tab)), 2)
      cont_tab <- rbind(cont_tab, thresholds)
      scales[1,] <- c(1,1,1, max(cont_tab$LRplus_test), max(cont_tab$LRplus_sub))
      scales[2,] <- c(0,0,0, min(cont_tab$LRplus_test), min(cont_tab$LRplus_sub))
      scales <- as.data.frame(scales)
      dimnames(scales) <- list(c("Max","Min"), colnames(cont_tab))
      cont_tab_lim <- rbind(scales, cont_tab)
          par(mfrow = c(1, 2))      
          fmsb::radarchart(cont_tab_lim, cglcol="grey", maxmin = TRUE, pcol = COL, plty=linetypes )
          plot.new() 
          graphics::legend("center", legend = rownames(cont_tab), title = "Strategy", 
            col = COL, seg.len = 2, border = "transparent", pch = 16, lty = linetypes)

      cont_tab$db_group <- NULL
      subchunk_count <- rechunk(DT::datatable(cont_tab_raw, filter = 'top', rownames = TRUE, extensions = c('Buttons','ColReorder'),
                                    options = list( paging =TRUE, colReorder = TRUE, dom = 'lftBip',  buttons = c('copy', 'csv', 'excel'))),
                          counter = subchunk_count, 
                          chunk_options = "echo=FALSE, results = 'asis', warning = FALSE, message = FALSE")
    }     

```

### **Prediction scores for the significant pairs**


The pairs in the multiMiR prediction databases are assigned scores that represent confidence in the predicted miRNA-target interaction.
For each prediction database, the scores for the significant pairs obtained by each strategy were compared with the scores in the **background**.
The **background** score distribution is the prediction score distribution of all possible pairs that can be generated combining all input data.

The comparison was performed using one-tailed bootstrap T-tests. 
Lower p-values means that the significant pairs had higher scores in the database.
The final column of the table shows the combined p-values for all databases, integrated using Fishers method.


```{r score_comp_pval, results = 'asis', echo = FALSE, warning = FALSE, message = FALSE, eval = FALSE}
# cat("

#   #### **One-tailed T-test**

#   ")
# score_comp_pval <- reshape(score_comp[,c("strategy","database","p.value")], idvar = "strategy", timevar = "database", direction = "wide")
#     colnames(score_comp_pval) <- gsub("p.value.", "",colnames(score_comp_pval))
#    DT::datatable(score_comp_pval, filter = 'top', rownames = FALSE, extensions = c('Buttons','ColReorder'),
#             options = list(
#               paging = FALSE,
#               colReorder = TRUE,
#               dom = 'lftBip',
#               buttons = c('copy', 'csv', 'excel')
#           ))

```


```{r score_comp_bootpval, results = 'asis', echo = FALSE, warning = FALSE, message = FALSE}
  score_comp_boot_pval <- reshape(score_comp[,c("strategy","database","boot.p.value")], idvar = "strategy", timevar = "database", direction = "wide")
   colnames(score_comp_boot_pval) <- gsub("boot.p.value.", "",colnames(score_comp_boot_pval))
  DT::datatable(score_comp_boot_pval, filter = 'top', rownames = TRUE, extensions = c('Buttons','ColReorder'),
            options = list(
              paging = FALSE,
              colReorder = TRUE,
              dom = 'lftBip',
              buttons = c('copy', 'csv', 'excel')
          ))

```


```{r score_comp_weighted_fisher, results = 'asis', echo = FALSE, warning = FALSE, message = FALSE, eval = FALSE}
#### **Weighted Fisher test**
  # score_comp_weighted_fisher <- reshape(score_comp[,c("strategy","database","w.fisher.p.value")], idvar = "strategy", timevar = "database", direction = "wide")
  #  colnames(score_comp_weighted_fisher) <- gsub("w.fisher.p.value.", "",colnames(score_comp_weighted_fisher))
  # DT::datatable(score_comp_weighted_fisher, filter = 'top', rownames = TRUE, extensions = c('Buttons','ColReorder'),
  #           options = list(
  #             paging = FALSE,
  #             colReorder = TRUE,
  #             dom = 'lftBip',
  #             buttons = c('copy', 'csv', 'excel')
  #         ))
```

#### **Comparison of p-values obtained in boostrap T-test**

This plot shows -log(p) values from the T-test and bootstrap T-test. Grey dotted line indicates -log10(0.05). **comb_pval** is the combination of all p-values using Fisher combined probability test.

```{r score_comp_pval_log, results = 'asis', echo = FALSE, warning = FALSE, message = FALSE}
  log_score_comp <- score_comp[,c("strategy","database","log.boot.p.value")]
# log_score_comp <- log_score_comp[!log_score_comp$strategy %in% unsig_strategies,]
    data.table::setnames(log_score_comp, "log.boot.p.value", "logPval")
   
  
  ggplot2::ggplot(log_score_comp, ggplot2::aes(x = database, y = logPval, color = strategy, group = strategy)) +
  ggplot2::geom_line() + 
  ggplot2::geom_line(ggplot2::aes( y =  -log10(0.05), linetype = "-log(0.05)" ), colour = "grey") +
  ggplot2::labs(colour = "Strategy", group = "Strategy", y = "-log(p)", linetype = "Threshold", y = "Database") +
  ggplot2::scale_linetype_manual(values=c("dashed", "solid")) #+
  # ggplot2::theme(legend.position = "bottom", legend.box = "vertical") + 
  # ggplot2::guides(color = ggplot2::guide_legend(title.position = "top"))
```


## **Strategies specific results**

```{r render_partial_reports, echo = FALSE, warning = FALSE}
  curr_out <- ""

  curr_out <- unlist(lapply(sig_strategies[sig_strategies != "DEGs_DEMs_permutated"] ,function(method_name){
      sec_text <- paste(c("### **", method_name ," correlation method**\n"),collapse="") # Header
      sec_text <- c(sec_text, knitr::knit_expand("partial/miRNA_cor_methods.Rmd")) # Body
      return(sec_text)
  }))    

``` 
`r paste(knitr::knit(text = curr_out), collapse = '\n')`


## **Fisher exact test significance (only for prediction databases)** 

Significant and non-significant pairs were compared in terms of overlap with databases using one-tailed Fisher exact tests. 
P-values are shown for the different strategies and for the different databases considered. 
-log10(p) are represented. The -log10(0.05) threshold is represented as a dashed grey line. **comb_pval** is the combination of all p-values using Fisher combined probability test.

```{r p_fisher, echo=FALSE, results = 'asis', warning = FALSE, message = FALSE}
    

    data.table::setnames(p_fisher, c("fisher.log.p.value", "fisher.p.value"), c("LogPval", "Pvalue"))
     ggplot2::ggplot(p_fisher, ggplot2::aes(x = database, y = LogPval, group = strategy, color = strategy)) +
    ggplot2::geom_line() + 
    ggplot2::geom_line(ggplot2::aes( y =  -log10(0.05), linetype = "-log10(0.05)") , colour = "grey") +
    ggplot2::geom_point(size = 1) +
    ggplot2::labs(colour = "Strategy", linetype = "-log10(0.05)", y = "-log10(p)", x = "Database") +
    ggplot2::scale_linetype_manual(values=c("dashed", "solid"))+
    ggplot2::theme(legend.position = "bottom", legend.box = "vertical")

```

```{r p_fisher_table, echo=FALSE, results = 'asis', warning = FALSE, message = FALSE}
    p_fisher <- p_fisher[, c("database", "strategy", "Pvalue")]
  p_fisher <- tidyr::spread(p_fisher, database, Pvalue)
    DT::datatable(p_fisher, filter = 'top', rownames = FALSE, extensions = c('Buttons','ColReorder'),
            options = list(
              paging = TRUE,
              colReorder = TRUE,
              dom = 'lftBip',
              buttons = c('copy', 'csv', 'excel')
          ))
```

## **Score distribution sampling check**
These plots compare the whole score distribution of background pairs with his subset. 

```{r score_sampling, results = "asis", echo = FALSE, warning = FALSE}
  for (database in selected_predicted_databases) {
    raw_score <- data.frame(x = raw_databases_scores[[database]])
    data.table::setnames(raw_score, "x", database)
    background_score <- all_pairs[,get(database)]
    background_score <- background_score[!is.na(background_score)]
    
    background_score <- data.frame(background_score)
    colnames(background_score) <- database

    gridExtra::grid.arrange (
    
      ggplot2::ggplot(raw_score,        ggplot2::aes_(x = as.name(database))) + 
                                               ggplot2::geom_density() + 
                                               ggplot2::ggtitle("Complete score distribution")
                                          ,
      ggplot2::ggplot(background_score, ggplot2::aes_(x = as.name(database))) + 
                                               ggplot2::geom_density() + 
                                               ggplot2::ggtitle("Background score distribution"),
      ncol = 2,
      top = grid::textGrob(stringr::str_to_title(database),gp=grid::gpar(fontsize=20,font=3)))
       
    }

  cat("\n\n")
```

### **Values of options used to run CorMiT**
First column contains the option names; second column contains the given values for each option in this run.
```{r opt_vals, echo = FALSE, warning = FALSE}
knitr::kable(cbind(opt))
```




